Multiple, configurable perceiver implementations, including with flash attention.

alxmrs opened the pull request:
This PR extracts the Perceiver into its own config. This allows us to specify which perceiver implementation to adopt in either the encoder or decoder (just the encoder for now). We add a flash attention perceiver as the second option beyond a "naive" perceiver. Last, users can select "auto" as an implementation option, and the config will pick the best implementation given the runtime environment.

alxmrs committed:
Multiple, configurable perceiver implementations, including with flash attention.

This PR extracts the Perceiver into its own config. This allows us to specify which perceiver implementation to adopt in either the encoder or decoder (just the encoder for now). We add a flash attention perceiver as the second option beyond a "naive" perceiver.

# Conflicts:
#	configs/train_om4_fomo.yaml
#	src/ocean_emulators/config.py
#	src/ocean_emulators/models/modules/encoder.py
#	tests/test_encoder.py

alxmrs committed:
Better error.

alxmrs committed:
Extra for cuda

alxmrs committed:
Ensure cuda deps are installed at deploy time.

alxmrs committed:
rm whitespace.

adam-subel commented on pyproject.toml:
```diff
@@ -27,7 +27,7 @@ dependencies = [
   "dacite>=1.9.2",
   "dask>=2025.2",
-  "einops>=0.8.1",
+  "einops",
```
I am not familiar with the choice in dependencies, but just wanted to note that this loosens a requirement and wanted to check that was intentional.

jder replied:
Yeah, is this conflicting with the new perceiver impl somehow? If so would be good to look into why this requirement was here in the first place.

alxmrs replied:
I'm happy to pin the requirement now that I loosened it; I did at one point need einops, I remember adding this dep.

adam-subel commented on src/ocean_emulators/config.py:
```diff
@@ -287,16 +295,61 @@ def build(
+            case "naive":
+                perceiver = NaivePerceiver(
+                    num_freq_bands=4,
+                    max_freq=1.0,
```
I notice that this is a departure from the previous code you shared for review. You had changes max_freq to be dependent on the encoder patch size. This is dropped now. Is that deemed to be uninportant?

adam-subel replied:
I suppose similar comments on num_latents. Also, maybe I am mixing the order of changes across these two commits.

adam-subel reviewed (approved):
Looks good. I left a couple comments I think maybe based more on my own confusion on how this interacts with the previous pull request you shared.

jder commented on skypilot/train.sky.yaml:
```diff
@@ -61,6 +61,9 @@ workdir: .
   set -ex
 
+  # Ensure CUDA dependencies are installed
+  uv sync --extra cuda
```
Should probably mention this somewhere in the docs, too?

alxmrs replied:
Good shout.

alxmrs committed:
Updated PerceiverConfig to incorporate encoder changes in main.

jder commented on pyproject.toml:
```diff
@@ -77,6 +77,10 @@ dev = [
+cuda = [
+  "flash-attn>=2.2.5,<3",
+  "flash-perceiver>=0.2",
+]
```
🐑 Not that it really matters right now, but I think these should probably be a set of [optional dependencies](https://docs.astral.sh/uv/concepts/projects/dependencies/#optional-dependencies) rather than a [development dependency group](https://docs.astral.sh/uv/concepts/projects/dependencies/#dependency-groups) since presumably people that pip install our package would want to run on GPUs.

jder commented on pyproject.toml:
```diff
@@ -117,3 +121,6 @@ check_untyped_defs = true
 follow_untyped_imports = true
+
+[tool.uv]
+no-build-isolation-package = [ 'flash-attn' ]
```
Rather than turning of build isolation completely, I suggest we do what the [uv docs give as an example for this package](https://docs.astral.sh/uv/concepts/projects/config/#augmenting-build-dependencies)

```suggestion
[tool.uv.extra-build-dependencies]
flash-attn = [{ requirement = "torch", match-runtime = true }]

[tool.uv.extra-build-variables]
flash-attn = { FLASH_ATTENTION_SKIP_CUDA_BUILD = "TRUE" }
```

jder commented on src/ocean_emulators/models/modules/encoder.py:
```diff
@@ -44,23 +39,9 @@ def __init__(
-        )
-        self.project = nn.Linear(perceiver_latent_dim, out_channels)
+        self.perceiver = perceiver
+        latent_dim: int = getattr(self.perceiver, "latent_dim")
```
This (and the get/set/hasattrs) seem like a smell to me. I would vote in favor of continuing to pass in this information separately, perhaps something like perceiver_output_shape? Alternatively you could include the final mean + projection into the passed nn.Module before passing it in here.

alxmrs replied:
This is the thing I was going to bring up with you! At the point where we're including the mean and projection in a Module we write, why wouldn't we simply just configure each implementation to do just that? They both support the same thing that we're implementing IIUC -- except, the naive flash attention mechanism also throws in a layer norm.

alxmrs replied:
I agree that the set/get attrs is less than ideal. I'll play around with it to see if there is another clean approach (will see what your suggestions look like in code).

alxmrs replied:
I now have an impl that is simpler, but it includes a LayerNorm instead of a simple linear projection. WDYT?

jder replied:
The reason I turned off the default behavior was to remove the LayerNorm. I don't really see why we'd want a LayerNorm there but I also don't have good evidence it's harmful.

jder commented on src/ocean_emulators/config.py:
```diff
@@ -287,31 +295,95 @@ def build(
+    def build(self, in_channels: int, patch_size: tuple[int, int]) -> nn.Module:
+        # This is not really a "frequency" but a maximum of the width appears to be reasonable from looking at the code.
+        max_freq = max(*patch_size)
+        match self.implementation:
```
Does this need to be configurable? Might be preferable to instead use the flash version iff we're using a CUDA backend and otherwise the non-flash one? (And yell if we're using a cuda backend but the flash implementation is not installed?) Put another way: what are we going to check in for this configuration value?

alxmrs replied:
I do think it should be configurable. While I have read each implementation and they look equivalent, the devil is in the (implementation) details. There may be a scientific reason why we would want to switch one implementation for another, outside of performance. Further, I have found yet another Perceiver implementation (that is well starred on GH), which we may want to experiment with later. Thus, I think this is worthy of a parameter that we explicitly control.

jder replied:
OK, then I think we should have an `auto` implementation which we use by default which does what I said above: if we're using a cuda backend, use "flash" (and error if it's not enabled) or else use "naive"

jder commented on src/ocean_emulators/config.py:
```diff
@@ -287,31 +295,95 @@ def build(
+            case "flash":
+                if not FLASH_ENABLED:
+                    raise ValueError(
+                        'The "flash" perceiver implementation does not work in this environment.'
```
```suggestion
                        'Configuration requested the flash perceiver, but the `flash_perceiver` package could not be imported'
```

jder commented on src/ocean_emulators/config.py:
```diff
@@ -287,31 +295,95 @@ def build(
+                    input_channels=in_channels,
+                    latent_dim=self.latent_dim,
+                    num_latents=self.num_latents,
+                    num_classes=self.latent_dim,  # This is ignored!
```
```suggestion
```

jder reviewed (changes requested):
Sorry for the delay in reviewing this!

alxmrs committed:
Adopting recommended way to use uv and flash attn

Co-authored-by: Jesse Rusak <jesse@openathena.ai>

alxmrs committed:
Better error msg.

Co-authored-by: Jesse Rusak <jesse@openathena.ai>

alxmrs committed:
Omit passing an argument that is ignored anyway.

Co-authored-by: Jesse Rusak <jesse@openathena.ai>

alxmrs committed:
Proper optional dep/extra + docs.

alxmrs committed:
Restricting einops dep

alxmrs committed:
Keeping things really simple -- configuring, not re-implementing.

This is a different impl than what was at before: both our Perceiver impls use a LayerNorm before linear projection. We may not want this, in which case, I will do something slightly more complex.

alxmrs committed:
Updated comment, added TODO.

alxmrs commented on skypilot/viz.sky.yaml:
```diff
@@ -62,6 +62,9 @@ workdir: .
   set -ex
 
+  # Ensure CUDA dependencies are installed
+  uv sync --extra cuda
```
@jder is CUDA needed in this case? I am now thinking: no.

jder replied:
Good point, yeah. I have actually been running this on AWS CPU nodes so would be nice to skip CUDA stuff.

jder commented on pyproject.toml:
```diff
@@ -27,7 +27,7 @@ dependencies = [
   "dacite>=1.9.2",
   "dask>=2025.2",
-  "einops>=0.8.1",
+  "einops>=0.7",
```
Why did this change from 0.8.1 to 0.7? Looks like the original requirements.txt included 0.8.0.

alxmrs replied:
I did it because of a version conflict with flash-perceivers: 

```
 × No solution found when resolving dependencies for split (python_full_version >= '3.13' and sys_platform == 'linux'):
  ╰─▶ Because only flash-perceiver<=0.2.0 is available and flash-perceiver==0.2.0 depends on einops>=0.7.0,<0.8.0, we can conclude that flash-perceiver>=0.2.0 depends on einops>=0.7.0,<0.8.0.
      And because ocean-emulators[cuda] depends on flash-perceiver>=0.2, we can conclude that ocean-emulators[cuda] depends on einops>=0.7.0,<0.8.0.
      And because your project depends on einops>=0.8 and your project requires ocean-emulators[cuda], we can conclude that your project's requirements are unsatisfiable
```

jder replied:
ok, skimming the release notes that seems ok. Might also be worth an upstream PR bumping to `0.7<=0.9` since I doubt 0.8 actually breaks anything.

jder commented on src/ocean_emulators/config.py:
```diff
@@ -287,31 +295,85 @@ def build(
+    ) -> nn.Module:
+        # This is not really a "frequency" but a maximum of the width appears to be reasonable from looking at the code.
+        max_freq = max(*patch_size)
+        # TODO(alxmrs,jder): Why compute the mean? Is it better to directly project from the num_latents x latent_dim?
```
Can you either clarify this is about the final output projection or rm the comment?

jder reviewed (approved):
Thanks, just a few follow-ups.

alxmrs committed:
No CUDA for Viz step.

alxmrs committed:
More in the TODO comment.

alxmrs committed:
Added an "auto" option to the PerceiverConfig

alxmrs committed:
docstring talks about auto.

alxmrs committed:
Bumping up the disk size.

jder commented on src/ocean_emulators/config.py:
```diff
@@ -287,31 +295,89 @@ def build(
+        #  Why compute the mean? Is it better to directly project from the num_latents x latent_dim?
+        if (
+            self.implementation == "auto" and FLASH_ENABLED
+        ) or self.implementation == "flash":
```
@alxmrs I think the behavior here is that if you forget to include `--extras cuda` it will silently fall back to the naive implementation. Can you please change this so that "auto" always means "flash" if we're using a torch cuda backend (ie so you get an error if you try to use it on GPU without installing the cuda extras)
